staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go

```diff
@@ -184,6 +184,11 @@ func WithMaxInFlightLimit(
 }
 
 func tooManyRequests(req *http.Request, w http.ResponseWriter) {
+	// Send a GOAWAY and tear down the TCP connection when idle.
+	// NOTE: the in-flight watching requests will NOT be influenced,
+	// and the new requests will use a new TCP connection and then load-balance will work,
+	// which is actually we want.
+	w.Header().Set("Connection", "close")
```

Comment:
cc @caesarxuchao @smarterclayton

Reply:
This needs to be probabilistic - tearing down every connection would be too expensive.  See the linked #48616 - I believe this needs to be much closer to that.

Reply:
I'd go further-- once the last apiserver becomes overloaded you end up with a complete system failure as every client request becomes drastically more expensive.

It's much, much better for us to queue requests. 1.18 has a useful alpha: https://github.com/kubernetes/kubernetes/pull/88177#issuecomment-588945806

I think best is for clients to occasionally do this, the client is actually in a better position to figure out if additional attempts have a good chance of finding a not-busy apiserver.

If we do this on the server, it should have a low probability, like 1/100.

We should never ever do this in a single-apiserver system.